comp/dogstatsd/server/batch.go

```diff
@@ -60,6 +63,53 @@ type batcher struct {
 	noAggPipelineEnabled bool
 }
 
+type shardKeyGenerator interface {
+	Generate(sample metrics.MetricSample, shards int) uint32
+}
+
+type shardKeyGeneratorBase struct {
+	keyGenerator *ckey.KeyGenerator
+	tagsBuffer   *tagset.HashingTagsAccumulator
+}
+
+func (s *shardKeyGeneratorBase) Generate(sample metrics.MetricSample, shards int) uint32 {
+	// TODO(remy): re-using this tagsBuffer later in the pipeline (by sharing
+	// it in the sample?) would reduce CPU usage, avoiding to recompute
+	// the tags hashes while generating the context key.
+	s.tagsBuffer.Append(sample.Tags...)
+	h := s.keyGenerator.Generate(sample.Name, sample.Host, s.tagsBuffer)
+	s.tagsBuffer.Reset()
+	return fastrange(h, shards)
+}
+
+type shardKeyGeneratorIsolation struct {
+	shardKeyGeneratorBase
+}
+
+func (s *shardKeyGeneratorIsolation) Generate(sample metrics.MetricSample, shards int) uint32 {
+	defer s.tagsBuffer.Reset()
+	for _, tag := range sample.Tags {
+		if strings.HasPrefix(tag, "service:") {
+			s.tagsBuffer.Append(tag)
+		}
+	}
+	// If we cannot isolate the samples based on the origin on known tags, we
+	// fallback on the base shard key generator.
+	if s.tagsBuffer.Len() == 0 && sample.OriginFromUDS == "" && sample.OriginFromClient == "" {
+		return s.shardKeyGeneratorBase.Generate(sample, shards)
+	}
+
+	// Otherwise, we isolate the samples based on the origin and well-known tags.
+
+	// WARNING: This might break aggregations if multiple containers send tags with different origins
```

Comment:
Aggregation across origins is quite widely used, I think we would need to support it even when isolation is used.